cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleDuringExecutionStore.java

```diff
@@ -0,0 +1,218 @@
+    producerProps.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
+    producerProps.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
+    producerProps.setProperty(ProducerConfig.RECONNECT_BACKOFF_MS_CONFIG, config.get(MonitorConfig.RECONNECT_BACKOFF_MS_CONFIG).toString());
+    return new KafkaProducer<>(producerProps);
+  }
+
+  @SuppressWarnings("unchecked")
+  protected void ensureTopicsCreated(Map<String, ?> config) {
+    AdminClient adminClient = KafkaCruiseControlUtils.createAdminClient((Map<String, Object>) config);
+    try {
+      short replicationFactor = sampleStoreTopicReplicationFactor(config, adminClient);
+
+      // Retention
+      long partitionSampleWindowMs = (Long) config.get(MonitorConfig.PARTITION_METRICS_WINDOW_MS_CONFIG);
+      int numPartitionSampleWindows = (Integer) config.get(MonitorConfig.NUM_PARTITION_METRICS_WINDOWS_CONFIG);
+      long partitionSampleRetentionMs = (numPartitionSampleWindows * ADDITIONAL_WINDOW_TO_RETAIN_FACTOR) * partitionSampleWindowMs;
+      partitionSampleRetentionMs = Math.max(_minPartitionSampleStoreTopicRetentionTimeMs, partitionSampleRetentionMs);
+
+      // New topics
+      if (_partitionMetricSampleStoreTopic != null) {
+        NewTopic partitionSampleStoreNewTopic =
+            wrapTopic(_partitionMetricSampleStoreTopic, _partitionSampleStoreTopicPartitionCount, replicationFactor,
+                partitionSampleRetentionMs);
+        ensureTopicCreated(adminClient, partitionSampleStoreNewTopic);
+      }
+
+    } finally {
+      KafkaCruiseControlUtils.closeAdminClientWithTimeout(adminClient);
+    }
+  }
+
+  protected void ensureTopicCreated(AdminClient adminClient, NewTopic sampleStoreTopic) {
+    if (!createTopic(adminClient, sampleStoreTopic)) {
+      // Update topic config and partition count to ensure desired properties.
+      maybeUpdateTopicConfig(adminClient, sampleStoreTopic);
+      maybeIncreasePartitionCount(adminClient, sampleStoreTopic);
+    }
+  }
+
+  @Override
+  public void storeSamples(MetricSampler.Samples samples) {
+    final AtomicInteger metricSampleCount = new AtomicInteger(0);
+    if (_partitionMetricSampleStoreTopic != null) {
+      for (PartitionMetricSample sample : samples.partitionMetricSamples()) {
+        _producer.send(new ProducerRecord<>(_partitionMetricSampleStoreTopic, null, sample.sampleTime(), null, sample.toBytes()),
+                       (recordMetadata, e) -> {
+                         if (e == null) {
+                           metricSampleCount.incrementAndGet();
+                         } else {
+                           LOG.error("Failed to produce partition metric sample for {} of timestamp {} due to exception",
+                               sample.entity().tp(), sample.sampleTime(), e);
+                         }
+                       });
+      }
+    }
```

Comment:
Duplicate code -- please see my comment on using abstract class to extract the common logic.